js/ui/overlay/ui.overlay.js

```diff
@@ -537,23 +560,32 @@ const Overlay = Widget.inherit({
         const hidingArgs = { cancel: false };
 
         if(this._isShowingActionCanceled) {
-            this._hidingDeferred.resolve();
+            delete this._isShowingActionCanceled;
+            this._hidingDeferred.reject();
         } else {
             this._actions.onHiding(hidingArgs);
 
             this._toggleSafariScrolling();
 
-            if(hidingArgs.cancel) {
+            const cancelHide = () => {
                 this._isHidingActionCanceled = true;
                 this.option('visible', true);
                 this._hidingDeferred.resolve();
-            } else {
+            };
+
+            const applyHide = () => {
                 this._forceFocusLost();
                 this._toggleShading(false);
                 this._toggleSubscriptions(false);
                 this._stopShowTimer();
 
                 this._animateHiding();
+            };
+
+            if(hidingArgs.cancel.then) {
+                hidingArgs.cancel.then(value => value ? cancelHide() : applyHide(), () => applyHide());
```

Comment:
Why so strange name, `value`? Let's make it more meaningful, `shouldCancel`, for example.

We should not use ternary operator if return value is not interesting to us.

Reply:
ok